### ISSUE TITLE
multiple master synchronization configuration infinite loop

### ISSUE THREAD

**Issue by silenceJI** [Icinga/icinga2#5794]
## Expected Behavior

master1 and master2 in "master cluster"

master1 through the API CRUD changes, if the API execution speed, for example, master1 configuration synchronization to master2, the API has been the master1 configuration is deleted, master2 again configuration synchronization to master1,  master2 received before master1 delete Configuration, will be synchronized again, resulting in an infinite loop;  If you modify the api.conf accept_config set to false, then back to normal, but this will result in each reload master1, master1 log will be a lot of refresh "ignore sync master2 configuration" In addition, this led to the master cluster is no longer fully available, if you use the API to connect master2 increase, delete, change, the configuration will no longer be synchronized to master1;

## Current Behavior

Same as above, problem time, master2 status:
![image](https://user-images.githubusercontent.com/19342102/33159686-6b2e0d36-d050-11e7-93fa-b73bd6dfc438.png)

last, master2 icinga process exit;

master1 log:

```
[2017-11-23 11:50:52 +0800] information/ConfigItem: Committing config item(s).
[2017-11-23 11:50:52 +0800] information/ConfigItem: Instantiated 1 Notification.
[2017-11-23 11:50:52 +0800] information/ConfigItem: Triggering Start signal for config items
[2017-11-23 11:50:52 +0800] information/ConfigItem: Activated all objects.
[2017-11-23 11:50:52 +0800] information/ConfigItem: Committing config item(s).
[2017-11-23 11:50:52 +0800] information/ConfigItem: Instantiated 1 Service.
[2017-11-23 11:50:52 +0800] information/ConfigItem: Triggering Start signal for config items
[2017-11-23 11:50:52 +0800] information/ConfigItem: Activated all objects.
[2017-11-23 11:50:52 +0800] information/ConfigItem: Committing config item(s).
[2017-11-23 11:50:52 +0800] information/ConfigItem: Instantiated 1 Notification.
[2017-11-23 11:50:52 +0800] information/ConfigItem: Triggering Start signal for config items
[2017-11-23 11:50:52 +0800] information/ConfigItem: Activated all objects.
[2017-11-23 11:50:52 +0800] information/ConfigItem: Committing config item(s).
[2017-11-23 11:50:52 +0800] information/ConfigItem: Instantiated 1 Notification.
[2017-11-23 11:50:52 +0800] information/ConfigItem: Triggering Start signal for config items
[2017-11-23 11:50:52 +0800] information/ConfigItem: Activated all objects.
[2017-11-23 11:50:53 +0800] information/ConfigItem: Committing config item(s).
[2017-11-23 11:50:53 +0800] information/ConfigItem: Instantiated 1 Notification.
[2017-11-23 11:50:53 +0800] information/ConfigItem: Triggering Start signal for config items
[2017-11-23 11:50:53 +0800] information/ConfigItem: Activated all objects.
[2017-11-23 11:50:53 +0800] information/ConfigItem: Committing config item(s).
[2017-11-23 11:50:53 +0800] information/ConfigItem: Instantiated 1 Notification.
[2017-11-23 11:50:53 +0800] information/ConfigItem: Triggering Start signal for config items
[2017-11-23 11:50:53 +0800] information/ConfigItem: Activated all objects.
[2017-11-23 11:50:53 +0800] information/ConfigItem: Committing config item(s).
[2017-11-23 11:50:53 +0800] information/ConfigItem: Instantiated 1 Notification.
[2017-11-23 11:50:53 +0800] information/ConfigItem: Triggering Start signal for config items
[2017-11-23 11:50:53 +0800] information/ConfigItem: Activated all objects.
[2017-11-23 11:50:53 +0800] information/ConfigItem: Committing config item(s).
[2017-11-23 11:50:53 +0800] information/ConfigItem: Instantiated 1 Notification.
[2017-11-23 11:50:53 +0800] information/ConfigItem: Triggering Start signal for config items
[2017-11-23 11:50:53 +0800] information/ConfigItem: Activated all objects.
[2017-11-23 11:50:53 +0800] information/ConfigItem: Committing config item(s).
[2017-11-23 11:50:53 +0800] information/ConfigItem: Instantiated 1 Notification.
[2017-11-23 11:50:53 +0800] information/ConfigItem: Triggering Start signal for config items
[2017-11-23 11:50:53 +0800] information/ConfigItem: Activated all objects.
[2017-11-23 11:50:54 +0800] information/ConfigItem: Committing config item(s).
[2017-11-23 11:50:54 +0800] information/ConfigItem: Instantiated 1 Notification.
[2017-11-23 11:50:54 +0800] information/ConfigItem: Triggering Start signal for config items
[2017-11-23 11:50:54 +0800] information/ConfigItem: Activated all objects.
[2017-11-23 11:50:54 +0800] information/ConfigItem: Committing config item(s).
[2017-11-23 11:50:54 +0800] information/ConfigItem: Instantiated 1 Notification.
[2017-11-23 11:50:54 +0800] information/ConfigItem: Triggering Start signal for config items
[2017-11-23 11:50:54 +0800] information/ConfigItem: Activated all objects.
[2017-11-23 11:50:54 +0800] information/ConfigItem: Committing config item(s).
[2017-11-23 11:50:54 +0800] information/ConfigItem: Instantiated 1 Notification.
[2017-11-23 11:50:54 +0800] information/ConfigItem: Triggering Start signal for config items
[2017-11-23 11:50:54 +0800] information/ConfigItem: Activated all objects.
[2017-11-23 11:50:54 +0800] information/ConfigItem: Committing config item(s).
[2017-11-23 11:50:54 +0800] information/ConfigItem: Instantiated 1 Notification.
[2017-11-23 11:50:54 +0800] information/ConfigItem: Triggering Start signal for config items
[2017-11-23 11:50:54 +0800] information/ConfigItem: Activated all objects.
[2017-11-23 11:50:54 +0800] information/ConfigItem: Committing config item(s).
[2017-11-23 11:50:54 +0800] information/ConfigItem: Instantiated 1 Notification.
[2017-11-23 11:50:54 +0800] information/ConfigItem: Triggering Start signal for config items
```

master2

```
[2017-11-23 11:50:53 +0800] information/ConfigItem: Committing config item(s).
[2017-11-23 11:50:53 +0800] information/ConfigItem: Instantiated 1 Notification.
[2017-11-23 11:50:53 +0800] information/ConfigItem: Triggering Start signal for config items
[2017-11-23 11:50:53 +0800] information/ConfigItem: Activated all objects.
[2017-11-23 11:50:53 +0800] information/ConfigItem: Committing config item(s).
[2017-11-23 11:50:53 +0800] information/ConfigItem: Instantiated 1 Notification.
[2017-11-23 11:50:53 +0800] information/ConfigItem: Triggering Start signal for config items
[2017-11-23 11:50:53 +0800] information/ConfigItem: Activated all objects.
[2017-11-23 11:50:54 +0800] information/ConfigItem: Committing config item(s).
[2017-11-23 11:50:54 +0800] information/ConfigItem: Instantiated 1 Notification.
[2017-11-23 11:50:54 +0800] information/ConfigItem: Triggering Start signal for config items
[2017-11-23 11:50:54 +0800] information/ConfigItem: Activated all objects.
[2017-11-23 11:50:54 +0800] information/ConfigItem: Committing config item(s).
[2017-11-23 11:50:54 +0800] information/ConfigItem: Instantiated 1 Notification.
[2017-11-23 11:50:54 +0800] information/ConfigItem: Triggering Start signal for config items
[2017-11-23 11:50:54 +0800] information/ConfigItem: Activated all objects.
[2017-11-23 11:50:54 +0800] information/ConfigItem: Committing config item(s).
[2017-11-23 11:50:54 +0800] information/ConfigItem: Instantiated 1 Notification.
[2017-11-23 11:50:54 +0800] information/ConfigItem: Triggering Start signal for config items
[2017-11-23 11:50:54 +0800] information/ConfigItem: Activated all objects.
[2017-11-23 11:50:54 +0800] information/ConfigItem: Committing config item(s).
[2017-11-23 11:50:54 +0800] information/ConfigItem: Instantiated 1 Notification.
[2017-11-23 11:50:54 +0800] information/ConfigItem: Triggering Start signal for config items
[2017-11-23 11:50:54 +0800] information/ConfigItem: Activated all objects.
[2017-11-23 11:50:54 +0800] information/ConfigItem: Committing config item(s).
[2017-11-23 11:50:54 +0800] information/ConfigItem: Instantiated 1 Notification.
[2017-11-23 11:50:54 +0800] information/ConfigItem: Triggering Start signal for config items
[2017-11-23 11:50:54 +0800] information/ConfigItem: Activated all objects.
[2017-11-23 11:50:55 +0800] information/ConfigItem: Committing config item(s).
[2017-11-23 11:50:55 +0800] information/ConfigItem: Instantiated 1 Notification.
[2017-11-23 11:50:55 +0800] information/ConfigItem: Triggering Start signal for config items
[2017-11-23 11:50:55 +0800] information/ConfigItem: Activated all objects.
[2017-11-23 11:50:55 +0800] information/ConfigItem: Committing config item(s).
[2017-11-23 11:50:55 +0800] information/ConfigItem: Instantiated 1 Notification.
[2017-11-23 11:50:55 +0800] information/ConfigItem: Triggering Start signal for config items
[2017-11-23 11:50:55 +0800] information/ConfigItem: Activated all objects.
[2017-11-23 11:50:55 +0800] information/ConfigItem: Committing config item(s).
[2017-11-23 11:50:55 +0800] information/ConfigItem: Instantiated 1 Service.
[2017-11-23 11:50:55 +0800] information/ConfigItem: Triggering Start signal for config items
[2017-11-23 11:50:55 +0800] information/ConfigItem: Activated all objects.
```

Executed API operation:

```
[2017-11-23 12:10:41 +0800] information/HttpServerConnection: Request: DELETE /v1/objects/notifications/l-fuwuticketi1.f.cn5%21check_ping%21push-notification (from [10.86.214.128]:51618, user: watcherapi)
[2017-11-23 12:10:41 +0800] information/HttpServerConnection: Request: DELETE /v1/objects/notifications/l-fuwuticketi1.f.cn5%21check_ping%21mail-notification (from [10.86.214.128]:51619, user: watcherapi)
[2017-11-23 12:10:41 +0800] information/HttpServerConnection: Request: DELETE /v1/objects/notifications/l-fuwuticketi1.f.cn5%21check_ping%21ivr-notification (from [10.86.214.128]:51620, user: watcherapi)
[2017-11-23 12:10:41 +0800] information/HttpServerConnection: Request: PUT /v1/objects/notifications/l-fuwuticketi1.f.cn5%21check_ping%21push-notification (from [10.86.214.128]:51621, user: watcherapi)
[2017-11-23 12:10:41 +0800] information/HttpServerConnection: Request: PUT /v1/objects/notifications/l-fuwuticketi1.f.cn5%21check_ping%21mail-notification (from [10.86.214.128]:51622, user: watcherapi)
[2017-11-23 12:10:41 +0800] information/HttpServerConnection: Request: PUT /v1/objects/notifications/l-fuwuticketi1.f.cn5%21check_ping%21ivr-notification (from [10.86.214.128]:51623, user: watcherapi)
[2017-11-23 12:10:42 +0800] information/HttpServerConnection: Request: DELETE /v1/objects/notifications/l-fuwuticketi3.f.cn5%21check_swap%21push-notification (from [10.86.214.128]:51633, user: watcherapi)
[2017-11-23 12:10:42 +0800] information/HttpServerConnection: Request: DELETE /v1/objects/notifications/l-fuwuticketi3.f.cn5%21check_swap%21mail-notification (from [10.86.214.128]:51634, user: watcherapi)
[2017-11-23 12:10:42 +0800] information/HttpServerConnection: Request: DELETE /v1/objects/notifications/l-fuwuticketi3.f.cn5%21check_swap%21ivr-notification (from [10.86.214.128]:51635, user: watcherapi)
[2017-11-23 12:10:42 +0800] information/HttpServerConnection: Request: PUT /v1/objects/notifications/l-fuwuticketi3.f.cn5%21check_swap%21push-notification (from [10.86.214.128]:51636, user: watcherapi)
[2017-11-23 12:10:43 +0800] information/HttpServerConnection: Request: PUT /v1/objects/notifications/l-fuwuticketi3.f.cn5%21check_swap%21mail-notification (from [10.86.214.128]:51637, user: watcherapi)
[2017-11-23 12:10:43 +0800] information/HttpServerConnection: Request: PUT /v1/objects/notifications/l-fuwuticketi3.f.cn5%21check_swap%21ivr-notification (from [10.86.214.128]:51638, user: watcherapi)
```

## Possible Solution

Modify master1 api.conf Configuration file: 

accept_config = false

## Steps to Reproduce (for bugs)

1.
2.
3.
4.

## Your Environment

* Version used (`icinga2 --version`):
master1 and master2:
```
icinga2 - The Icinga 2 network monitoring daemon (version: r2.7.1-1)

Copyright (c) 2012-2017 Icinga Development Team (https://www.icinga.com/)
License GPLv2+: GNU GPL version 2 or later <http://gnu.org/licenses/gpl2.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Application information:
  Installation root: /usr
  Sysconf directory: /etc
  Run directory: /run
  Local state directory: /var
  Package data directory: /usr/share/icinga2
  State path: /var/lib/icinga2/icinga2.state
  Modified attributes path: /var/lib/icinga2/modified-attributes.conf
  Objects path: /var/cache/icinga2/icinga2.debug
  Vars path: /var/cache/icinga2/icinga2.vars
  PID path: /run/icinga2/icinga2.pid

System information:
  Platform: CentOS Linux
  Platform version: 7 (Core)
  Kernel: Linux
  Kernel version: 3.10.0-327.36.3.el7.x86_64
  Architecture: x86_64

Build information:
  Compiler: GNU 4.8.5
  Build host: unknown
```
* Operating System and version:
master1 and master2
```
CentOS Linux release 7.2.1511 (Core)
```
* Enabled features (`icinga2 feature list`):
master1 and master2
```
Disabled features: command compatlog debuglog gelf graphite influxdb livestatus opentsdb perfdata statusdata syslog
Enabled features: api checker ido-mysql mainlog notification
```
* Icinga Web 2 version and modules (System - About):
* Config validation (`icinga2 daemon -C`):
```
information/cli: Icinga application loader (version: r2.7.1-1)
information/cli: Loading configuration file(s).
information/ConfigItem: Committing config item(s).
information/ApiListener: My API identity: l-qicingam1.ops.cn5
information/WorkQueue: #4 (DaemonUtility::LoadConfigFiles) items: 0, rate: 4216.58/s (252995/min 252995/5min 252995/15min);
information/WorkQueue: #5 (ApiListener, RelayQueue) items: 0, rate:  0/s (0/min 0/5min 0/15min);
information/WorkQueue: #6 (ApiListener, SyncQueue) items: 0, rate:  0/s (0/min 0/5min 0/15min);
information/WorkQueue: #7 (IdoMysqlConnection, ido-mysql) items: 0, rate:  0/s (0/min 0/5min 0/15min);
information/ConfigItem: Instantiated 5 ApiUsers.
information/ConfigItem: Instantiated 1 ApiListener.
information/ConfigItem: Instantiated 3 Zones.
information/ConfigItem: Instantiated 1 FileLogger.
information/ConfigItem: Instantiated 4 Endpoints.
information/ConfigItem: Instantiated 1 UserGroup.
information/ConfigItem: Instantiated 85421 Notifications.
information/ConfigItem: Instantiated 4 NotificationCommands.
information/ConfigItem: Instantiated 422 CheckCommands.
information/ConfigItem: Instantiated 121 Downtimes.
information/ConfigItem: Instantiated 2 HostGroups.
information/ConfigItem: Instantiated 1 IcingaApplication.
information/ConfigItem: Instantiated 1 EventCommand.
information/ConfigItem: Instantiated 4023 Hosts.
information/ConfigItem: Instantiated 10063 Users.
information/ConfigItem: Instantiated 3 TimePeriods.
information/ConfigItem: Instantiated 69181 Services.
information/ConfigItem: Instantiated 4 ServiceGroups.
information/ConfigItem: Instantiated 1 CheckerComponent.
information/ConfigItem: Instantiated 1 IdoMysqlConnection.
information/ConfigItem: Instantiated 1 NotificationComponent.
information/ScriptGlobal: Dumping variables to file '/var/cache/icinga2/icinga2.vars'
information/cli: Finished validating the configuration file(s).
```
* If you run multiple Icinga 2 instances, the `zones.conf` file (or `icinga2 object list --type Endpoint` and `icinga2 object list --type Zone`) from all affected nodes.
master1 and master2
```
/*
 * Generated by Icinga 2 node setup commands
 * on 2017-01-09 17:12:48 +0800
 */

object Endpoint "l-qicingam1.ops.cn5" {
	host = "10.88.197.234"
//	log_duration = 5m
}

object Endpoint "l-qicingam2.ops.cn5" {
	host = "10.88.197.235"
//	log_duration = 5m
}

object Endpoint "l-qicingan1.ops.cn5" {
	host = "10.88.128.52"
	log_duration = 0
}

object Endpoint "l-qicingan2.ops.cn5" {
	host = "10.88.128.19"
	log_duration = 0
}

object Zone "master" {
	endpoints = [ "l-qicingam1.ops.cn5", "l-qicingam2.ops.cn5" ]
}

object Zone "zone-cn5" {
	endpoints = [ "l-qicingan1.ops.cn5", "l-qicingan2.ops.cn5" ]
	parent = "master"
}

/* sync global commands */
object Zone "global-templates" {
        global = true
}
```

**Comment by silenceJI** [comment]
HI, 

Problem update

  Because of our large cluster, carrying 70,000 monitoring points, more than 80,000 notifications; a total of 170,000 objects;

  When we set MASTER-1 api.conf to `accept_config = false`;

1.  Through the API (delete, add) a service object, sometimes appear:
```
 {"results": [{"status": "Object could not be created.", "code": 500.0, "errors": ["Configuration file" / var / lib / icinga2 / api / packages / _api / cn0qicingam1.ops.cn2-1499764333-1 / conf.d / notifications / l-likunti4.dba.dev.cn0! check_load! push-notification.conf 'already exists. "]}]}
```

  Object already exists, and then we manually check the object through the API, there will be 404;

```
curl -k -s -u xxx: xxx-H 'Accpet: application / json' -X GET 'https://10.90.0.81:5665/v1/objects/notifications/l-likunti4.dba.dev.cn0!check_load ! push-notification '| jq
{
  "error": 404,
  "status": "No objects found."
}
```

  Reload MASTER-1 after recovery;

2. After each reload, master1 will receive 170,000 times the synchronization configuration update sent by master2, because set `accept_config = false`, each synchronization request will have a warning to the log;

```
[2017-12-15 10:20:25 +0800] warning/ApiListener: Ignoring config update from 'l-qicingam4.ops.cn5' for object 'l-advertisement14.wap.cn5' of type 'Host'. 'api' does not accept config.
[2017-12-15 10:20:25 +0800] warning/ApiListener: Ignoring config update from 'l-qicingam4.ops.cn5' for object 'l-advertisement15.wap.cn5' of type 'Host'. 'api' does not accept config.
[2017-12-15 10:20:25 +0800] warning/ApiListener: Ignoring config update from 'l-qicingam4.ops.cn5' for object 'l-advertisement22.wap.cn5' of type 'Host'. 'api' does not accept config.
[2017-12-15 10:20:25 +0800] warning/ApiListener: Ignoring config update from 'l-qicingam4.ops.cn5' for object 'l-advertisement13.wap.cn5' of type 'Host'. 'api' does not accept config.
[2017-12-15 10:20:25 +0800] warning/ApiListener: Ignoring config update from 'l-qicingam4.ops.cn5' for object 'l-advertisement18.wap.cn5' of type 'Host'. 'api' does not accept config.
[2017-12-15 10:20:25 +0800] warning/ApiListener: Ignoring config update from 'l-qicingam4.ops.cn5' for object 'l-advertisement3.wap.cn5' of type 'Host'. 'api' does not accept config.
[2017-12-15 10:20:25 +0800] warning/ApiListener: Ignoring config update from 'l-qicingam4.ops.cn5' for object 'l-advertisement21.wap.cn5' of type 'Host'. 'api' does not accept config.
[2017-12-15 10:20:25 +0800] warning/ApiListener: Ignoring config update from 'l-qicingam4.ops.cn5' for object 'l-advertisement1.wap.cn5' of type 'Host'. 'api' does not accept config.
[2017-12-15 10:20:25 +0800] warning/ApiListener: Ignoring config update from 'l-qicingam4.ops.cn5' for object 'l-advertisement25.wap.cn5' of type 'Host'. 'api' does not accept config.
[2017-12-15 10:20:25 +0800] warning/ApiListener: Ignoring config update from 'l-qicingam4.ops.cn5' for object 'l-advertisement23.wap.cn5' of type 'Host'. 'api' does not accept config.
[2017-12-15 10:20:25 +0800] warning/ApiListener: Ignoring config update from 'l-qicingam4.ops.cn5' for object 'l-advertisement24.wap.cn5' of type 'Host'. 'api' does not accept config.
```

3. Set 500 hosts downtime via ICINGA API; number of concurrent calls 4;

  MASTER1 Spent 5 minutes to process the request, MASTER2 sync these downtime objects for nearly 15 minutes;

Master2 : 
![image](https://user-images.githubusercontent.com/19342102/34040041-097c98b2-e1cd-11e7-816e-a3f1a0208e98.png)


  Checker node synchronization takes about 6 minutes;

MASTER cluster internal synchronization mechanism should have a problem;

**Comment by Al2Klimov** [comment]
I'm afraid that's exactly the problem our customers have with ScheduledDowntimes – they create config objects exactly like the API.

**Comment by Al2Klimov** [comment]
Just discovered that `ConfigObjectUtility#CreateObject()` first check whether an object exists and if not, it creates it. This may lead to race conditions especially in large setups.

Dear core devs, what about a general R/W lock per config object type? Or would that throw away too much performance?

**Comment by ekeih** [comment]
I have trouble to understand the initial issue description but I think @Al2Klimov is right.
At least we also often have a bunch of `Cannot create object ... already exists` messages in our log during the cluster synchronisation.

**Comment by Al2Klimov** [comment]
@dnsmichi IMO we should talk next year (offline) about the pandora cans in config+cluster and how to close them as smart as possible.

**Comment by dnsmichi** [comment]
The underlaying config package `_api` is broken which renders this ping-pong behaviour. I've fixed this for 2.11 with #7150.
